nsxt/metadata/metadata.go

```diff
@@ -0,0 +1,256 @@
+				} else {
+					d.Set(key, nestedSlice)
+				}
+			}
+		} else {
+			if len(parent) > 0 {
+				log.Printf("[INFO] assigning nested value %+v to %s", elem.FieldByName(item.Metadata.SdkFieldName).Interface(), key)
+				parentMap[key] = elem.FieldByName(item.Metadata.SdkFieldName).Interface()
+			} else {
+				log.Printf("[INFO] assigning value %+v to %s", elem.FieldByName(item.Metadata.SdkFieldName).Interface(), key)
+				d.Set(key, elem.FieldByName(item.Metadata.SdkFieldName).Interface())
+			}
+		}
+	}
+}
+
+// SchemaToStruct converts terraform schema to NSX model struct
+// currently supports nested subtype and trivial types
+func SchemaToStruct(elem reflect.Value, d *schema.ResourceData, metadata map[string]*ExtendedSchema, parent string, parentMap map[string]interface{}) {
+	for key, item := range metadata {
+		if item.Metadata.ReadOnly || item.Metadata.Skip {
+			continue
+		}
+		if item.Metadata.IntroducedInVersion != "" && util.NsxVersionLower(item.Metadata.IntroducedInVersion) {
+			continue
+		}
+
+		log.Printf("[INFO] inspecting key %s with type %s", key, item.Metadata.SchemaType)
+		if len(parent) > 0 {
+			log.Printf("[INFO] parent %s key %s", parent, key)
+		}
+		if item.Metadata.SchemaType == "string" {
+			var value string
+			if len(parent) > 0 {
+				value = parentMap[key].(string)
+			} else {
+				value = d.Get(key).(string)
+			}
+			log.Printf("[INFO] assigning string %v to %s", value, key)
+			elem.FieldByName(item.Metadata.SdkFieldName).Set(reflect.ValueOf(&value))
+		}
+		if item.Metadata.SchemaType == "bool" {
+			var value bool
+			if len(parent) > 0 {
+				value = parentMap[key].(bool)
+			} else {
+				value = d.Get(key).(bool)
+			}
+			log.Printf("[INFO] assigning bool %v to %s", value, key)
+			elem.FieldByName(item.Metadata.SdkFieldName).Set(reflect.ValueOf(&value))
+		}
+		if item.Metadata.SchemaType == "int" {
+			var value int64
+			if len(parent) > 0 {
+				value = int64(parentMap[key].(int))
+			} else {
+				value = int64(d.Get(key).(int))
+			}
+			log.Printf("[INFO] assigning int %v to %s", value, key)
+			elem.FieldByName(item.Metadata.SdkFieldName).Set(reflect.ValueOf(&value))
+		}
+		if item.Metadata.SchemaType == "struct" {
+			nestedObj := reflect.New(item.Metadata.ReflectType)
+			var itemList []interface{}
+			if len(parent) > 0 {
+				itemList = parentMap[key].([]interface{})
+			} else {
+				itemList = d.Get(key).([]interface{})
+			}
+			if len(itemList) == 0 {
+				continue
+			}
+			nestedSchema := itemList[0].(map[string]interface{})
+
+			childElem := item.Schema.Elem.(*ExtendedResource)
+			SchemaToStruct(nestedObj.Elem(), d, childElem.Schema, key, nestedSchema)
+			log.Printf("[INFO] assigning struct %v to %s", nestedObj, key)
+			elem.FieldByName(item.Metadata.SdkFieldName).Set(nestedObj)
+		}
+		if item.Metadata.SchemaType == "list" || item.Metadata.SchemaType == "set" {
+			var itemList []interface{}
+			if len(parent) > 0 {
+				itemList = parentMap[key].([]interface{})
```

Comment:
I don't think this conversion can work with `TypeSet` (f.e., to convert `TypeSet` to string list, we use `d.Get(schemaAttrName).(*schema.Set).List()`
Is there an example of `TypeSet` within this PR?

Reply:
Fixed and added in UT